File: Python/Arrays/pg0003.py
#Rotate array by k – Shift all elements right or left by k positions.

n = int(input())
arr = list(map(int,input().split()))
k = int(input())
k = k % n
arr = arr[-k:] + arr[:-k]
print(arr)

"""
# Alternatively, using reversal algorithm
def reverse(sub_arr, start, end):
    while start < end:
        sub_arr[start], sub_arr[end] = sub_arr[end], sub_arr[start]
        start += 1
        end -= 1
reverse(arr, 0, n-1)
reverse(arr, 0, k-1)
reverse(arr, k, n-1)
print(arr)
"""

"""
# Using extra space
temp = arr[-k:]
for i in range(n-k-1, -1, -1):
    arr[i+k] = arr[i]
for i in range(k):
    arr[i] = temp[i]
print(arr)
"""

File: Java/Arrays/pg0003.java
//Rotate array by k – Shift all elements right or left by k positions.

import java.util.*;

class pg0003{
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        if(n==0){
            System.out.println("Array is empty");
            return;
        }
        int arr[] = new int[n];
        for(int i=0;i<n;i++){
            arr[i] = sc.nextInt();
        }
        int k = sc.nextInt();
        k = k % n;
        if(k<0){
            k += n;
        }
        reverse(arr, 0, n - 1);
        reverse(arr, 0, k - 1);
        reverse(arr, k, n - 1);
        for(int i:arr){
            System.out.print(i+" ");
        }
    }
    
    static void reverse(int arr[], int start, int end){
        while(start<end){
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }
    }
}
